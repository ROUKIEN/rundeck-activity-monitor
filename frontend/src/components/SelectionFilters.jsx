import React from 'react'
import DropdownFilter from './DropdownFilter'

const isChecked = (selection, property, value) =>
  property in selection === true && selection[property].includes(value)

const SelectionFilters = ({ filters, selection, onChange }) => {

  const handleChange = (field, value) => {
    if (field in selection === true) {
      if (selection[field].includes(value)) {
        selection[field] = selection[field].filter(item => item !== value)
        if (selection[field].length === 0) {
          delete selection[field]
        }
      } else {
        selection[field].push(value)
      }
    } else {
      selection[field] = [
        value
      ]
    }

    onChange(selection)
  }

  return (
    <>
      <p className="menu-label">
        Search
      </p>
      <ul className="menu-list">
        { Object.keys(filters).map(filter => {
          return (
            <DropdownFilter
              filter={filter}
              count={filters[filter].length}
              key={filter}>
              { filters[filter].map(value =>
                <li key={value}>
                  <label className="checkbox">
                    <input
                      key={Math.random()}
                      type="checkbox"
                      defaultChecked={isChecked(selection, filter, value)}
                      onChange={e => handleChange(filter, value, e)}
                    ></input>&nbsp;{value}
                  </label>
                </li>
              )}
            </DropdownFilter>
          )
        })}
        <li>
          <button
            className="button is-small is-primary is-fullwidth"
            data-testid="reset-search-btn"
            onClick={ () => {
              onChange({})
            }}
          >Reset Filters&nbsp;<span className="delete"></span></button>
        </li>
      </ul>
    </>
  )
}

export default SelectionFilters
