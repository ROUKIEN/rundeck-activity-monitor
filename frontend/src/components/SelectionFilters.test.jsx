import React from 'react'
import renderer from 'react-test-renderer'
import { expect, test, vi } from 'vitest'
import SelectionFilters from './SelectionFilters'

test('SelectionFilters renders correctly', () => {
  const filters = {
    instances: ["foo", "bar", "zee"],
    statuses: ["succeeded", "failed"]
  }
  const tc = renderer.create(
    <SelectionFilters filters={ filters } selection={ {} }></SelectionFilters>
  ).toJSON()
  expect(tc).toBeDefined()
  expect(tc).toMatchSnapshot('Rendered list without any facet selected')
})

test('SelectionFilters renders with selection', () => {
  const filters = {
    instances: ["foo", "bar", "zee"],
    statuses: ["succeeded", "failed"]
  }

  const selection = {
    instances: ["foo"]
  }

  const tc = renderer.create(
    <SelectionFilters filters={ filters } selection={ selection }></SelectionFilters>
  ).toJSON()
  expect(tc).toBeDefined()
  expect(tc).toMatchSnapshot('Rendered list with foo instance selected')
})

test('SelectionFilters triggers onChange when something gets selected', () => {
  const filters = {
    instances: ['foo', 'bar', 'zee'],
    statuses: ['succeeded', 'failed']
  }

  const selection = {
    instances: ['foo']
  }

  const fn = vi.fn()

  const tc = renderer.create(
    <SelectionFilters
      filters={ filters }
      selection={ selection }
      onChange={fn}
    ></SelectionFilters>
  )

  tc.root.findAllByType('input')[3].props.onChange()
  expect(fn).toHaveBeenCalledTimes(1)
  expect(fn).toHaveBeenCalledWith({
    instances: ['foo'],
    statuses: ['succeeded'],
  })
})

test('SelectionFilters triggers onChange with empty selection when reset filters button is clicked', () => {
  const filters = {
    instances: ['foo', 'bar', 'zee'],
    statuses: ['succeeded', 'failed']
  }

  const selection = {
    instances: ['foo', 'bar'],
    statuses: ['failed']
  }

  const fn = vi.fn()

  const tc = renderer.create(
    <SelectionFilters
      filters={ filters }
      selection={ selection }
      onChange={fn}
    ></SelectionFilters>
  )

  tc.root.findByType('button').props.onClick()
  expect(fn).toHaveBeenCalledTimes(1)
  expect(fn).toHaveBeenCalledWith({})
})

test('SelectionFilters onChange triggers unselected ', () => {
  const filters = {
    instances: ['foo', 'bar', 'zee'],
    statuses: ['succeeded', 'failed']
  }

  const selection = {
    instances: ['foo', 'bar'],
    statuses: ['failed']
  }

  const fn = vi.fn()

  const tc = renderer.create(
    <SelectionFilters
      filters={ filters }
      selection={ selection }
      onChange={fn}
    ></SelectionFilters>
  )

  tc.root.findAllByType('input')[0].props.onChange()
  expect(fn).toHaveBeenCalledTimes(1)
  expect(fn).toHaveBeenCalledWith({
    instances: ['bar'],
    statuses: ['failed']
  })
})
