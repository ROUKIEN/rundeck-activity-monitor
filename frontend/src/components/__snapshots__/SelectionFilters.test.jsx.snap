// Vitest Snapshot v1

exports[`SelectionFilters renders correctly > Rendered list without any facet selected 1`] = `
[
  <p
    className="menu-label"
  >
    Search
  </p>,
  <ul
    className="menu-list"
  >
    <li>
      <button
        className="button is-small is-fullwidth"
        onClick={[Function]}
        style={
          {
            "cursor": "pointer",
          }
        }
      >
        <span
          className="is-capitalized"
        >
          instances
           (
          3
          )
        </span>
        <expandIcon
          expanded={true}
        />
      </button>
      <ul>
        <li>
          <label
            className="checkbox"
          >
            <input
              defaultChecked={false}
              onChange={[Function]}
              type="checkbox"
            />
             
            foo
          </label>
        </li>
        <li>
          <label
            className="checkbox"
          >
            <input
              defaultChecked={false}
              onChange={[Function]}
              type="checkbox"
            />
             
            bar
          </label>
        </li>
        <li>
          <label
            className="checkbox"
          >
            <input
              defaultChecked={false}
              onChange={[Function]}
              type="checkbox"
            />
             
            zee
          </label>
        </li>
      </ul>
    </li>
    <li>
      <button
        className="button is-small is-fullwidth"
        onClick={[Function]}
        style={
          {
            "cursor": "pointer",
          }
        }
      >
        <span
          className="is-capitalized"
        >
          statuses
           (
          2
          )
        </span>
        <expandIcon
          expanded={true}
        />
      </button>
      <ul>
        <li>
          <label
            className="checkbox"
          >
            <input
              defaultChecked={false}
              onChange={[Function]}
              type="checkbox"
            />
             
            succeeded
          </label>
        </li>
        <li>
          <label
            className="checkbox"
          >
            <input
              defaultChecked={false}
              onChange={[Function]}
              type="checkbox"
            />
             
            failed
          </label>
        </li>
      </ul>
    </li>
    <li>
      <button
        className="button is-small is-primary is-fullwidth"
        data-testid="reset-search-btn"
        onClick={[Function]}
      >
        Reset Filters 
        <span
          className="delete"
        />
      </button>
    </li>
  </ul>,
]
`;

exports[`SelectionFilters renders with selection > Rendered list with foo instance selected 1`] = `
[
  <p
    className="menu-label"
  >
    Search
  </p>,
  <ul
    className="menu-list"
  >
    <li>
      <button
        className="button is-small is-fullwidth"
        onClick={[Function]}
        style={
          {
            "cursor": "pointer",
          }
        }
      >
        <span
          className="is-capitalized"
        >
          instances
           (
          3
          )
        </span>
        <expandIcon
          expanded={true}
        />
      </button>
      <ul>
        <li>
          <label
            className="checkbox"
          >
            <input
              defaultChecked={true}
              onChange={[Function]}
              type="checkbox"
            />
             
            foo
          </label>
        </li>
        <li>
          <label
            className="checkbox"
          >
            <input
              defaultChecked={false}
              onChange={[Function]}
              type="checkbox"
            />
             
            bar
          </label>
        </li>
        <li>
          <label
            className="checkbox"
          >
            <input
              defaultChecked={false}
              onChange={[Function]}
              type="checkbox"
            />
             
            zee
          </label>
        </li>
      </ul>
    </li>
    <li>
      <button
        className="button is-small is-fullwidth"
        onClick={[Function]}
        style={
          {
            "cursor": "pointer",
          }
        }
      >
        <span
          className="is-capitalized"
        >
          statuses
           (
          2
          )
        </span>
        <expandIcon
          expanded={true}
        />
      </button>
      <ul>
        <li>
          <label
            className="checkbox"
          >
            <input
              defaultChecked={false}
              onChange={[Function]}
              type="checkbox"
            />
             
            succeeded
          </label>
        </li>
        <li>
          <label
            className="checkbox"
          >
            <input
              defaultChecked={false}
              onChange={[Function]}
              type="checkbox"
            />
             
            failed
          </label>
        </li>
      </ul>
    </li>
    <li>
      <button
        className="button is-small is-primary is-fullwidth"
        data-testid="reset-search-btn"
        onClick={[Function]}
      >
        Reset Filters 
        <span
          className="delete"
        />
      </button>
    </li>
  </ul>,
]
`;
